include/clang/AST/ASTImporterSharedState.h

```diff
@@ -16,18 +16,24 @@
 #define LLVM_CLANG_AST_ASTIMPORTERSHAREDSTATE_H
 
 #include "clang/AST/ASTImporterLookupTable.h"
+#include "clang/Basic/SourceLocation.h"
 #include "llvm/ADT/DenseMap.h"
 // FIXME We need this because of ImportError.
 #include "clang/AST/ASTImporter.h"
 
 namespace clang {
 
+class ASTUnit;
 class TranslationUnitDecl;
 
 /// Importer specific state, which may be shared amongst several ASTImporter
 /// objects.
 class ASTImporterSharedState {
+public:
+  using ImportedFileIDMap =
+      llvm::DenseMap<FileID, std::pair<FileID, ASTUnit *>>;
```

Comment:
Why must we store this info in the shared state? Could you please elaborate?

Reply:
The goal is to get the `ASTUnit` from a `FileID` that was imported, by only knowing the `FileID` value (not the `ASTImporter` or `ASTUnit`). This map can be used to get the ASTUnit if there are multiple ASTImporters (and corresponding ASTUnits). (Previously the filename of the source file was used to get the ASTUnit in `CrossTranslationUnitContext` but this does not work if the file is an include file. And using the filename can be source of errors.)

Reply:
Why this can't be a member of the ASTImporter class? Why would it be a problem if the different ASTImporter objects have their own copy?

Reply:
This map is needed to find out the `ASTUnit` in `CrossTranslationUnitContext::GetImportedFromSourceLocation`. It should contain `FileID`'s imported by any of the `ASTImporter`'s and the data is used independent of a single `ASTImporter`. Every `ASTImporter` inserts its own `ASTUnit` into it (this is why the `ASTUnit` is added to `ASTImporter`: The `ASTImporter` can get the `FileID` mapping and at the same time the `ASTUnit`). (If every `ASTImporter` had its own copy we can loop over every `ASTUnit` to find the matching one. This can be an alternative solution, then a simple FileID->FileID map is sufficient in `ASTImporter`, and no store of `ASTUnit`. But the lookup becomes slower.)

Reply:
Okay, could you please then add a comment to the member which describes this (also mentioning the alternative and why that would be slower).